keras_nlp/utils/pipeline_model.py

```diff
@@ -0,0 +1,149 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A base class for models including preprocessing."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+def convert_inputs_to_dataset(
+    x=None,
+    y=None,
+    sample_weight=None,
+    batch_size=None,
+):
+    if isinstance(x, tf.data.Dataset):
+        if y is not None:
+            raise ValueError(
+                "When `x` is a `tf.data.Dataset`, please do not provide "
+                f"`y`. Received: `y={y}`."
+            )
+        if sample_weight is not None:
+            raise ValueError(
+                "When `x` is a `tf.data.Dataset`, please do not provide "
+                f"`sample_weight`. Received: `sample_weight={sample_weight}`."
+            )
+        return x
+
+    inputs = keras.utils.pack_x_y_sample_weight(x, y, sample_weight)
+    return tf.data.Dataset.from_tensor_slices(inputs).batch(batch_size or 32)
+
+
+def apply_dataset_fn(ds, fn):
+    return ds.map(fn, tf.data.AUTOTUNE).prefetch(tf.data.AUTOTUNE)
+
+
+class PipelineModel(keras.Model):
+    """A model which allows automatically applying preprocessing."""
+
+    def __init__(self, *args, include_preprocessing=True, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.include_preprocessing = include_preprocessing
+
+    def preprocess_features(self, x):
+        """An overridable function which preprocesses features."""
+        return x
+
+    def preprocess_labels(self, y):
+        """An overridable function which preprocesses labels."""
+        return y
+
+    def preprocess(self, *args):
+        """An overridable function which preprocesses samples."""
+        x, y, sample_weight = keras.utils.unpack_x_y_sample_weight(args)
+        x, y = self.preprocess_features(x), self.preprocess_labels(y)
+        return keras.utils.pack_x_y_sample_weight(x, y, sample_weight)
+
+    # ========================================================================
+    # Below are overrides to keras.Model methods to apply the functions above.
+    # ========================================================================
+    def fit(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        validation_data=None,
+        validation_split=None,
+        **kwargs,
+    ):
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = apply_dataset_fn(x, self.preprocess)
+
+        if validation_split:
```

Comment:
We should support it. It's straightforward -- see RetinaNet example

Reply:
Ah yeah, I think there are some bugs in the CV version. We can add this, just we need a little more complexity than what's there.

[These lines](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/object_detection/object_detection_base_model.py#L47-L53) actually keep the validation data in the training data! Pretty broken I think.

Will ping the CV folk.